src/schema.rs

```diff
@@ -321,30 +342,51 @@ impl RecordField {
 pub struct UnionSchema {
     schemas: Vec<Schema>,
     // Used to ensure uniqueness of schema inputs, and provide constant time finding of the
-    // schema index given a value.
-    // **NOTE** that this approach does not work for named types, and will have to be modified
-    // to support that. A simple solution is to also keep a mapping of the names used.
+    // schema index given a name or value.
+    name_index: HashMap<String, usize>,
     variant_index: HashMap<SchemaKind, usize>,
```

Comment:
Crazy idea: what about trying to unify these two maps into one?
If we could use `(SchemaKind, Option<String>)` as key for the index, I think we could still have O(1) lookup in `find_schema`.

Reply:
That's not a bad idea. I'll try that.